### DISCUSSION TITLE
Home Assistant access with External NPM

### DISCUSSION THREAD

**Discussion by cschwarck** [NginxProxyManager/nginx-proxy-manager#1401] [Q&A]
Hello All - I'm attempting to access my Home Assistant system through NPM, but it's failing with a 404. After doing some research, it looks like I need to add this to work: 

Include proxy_set_header in both location blocks: / and /api/websocket. 
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "upgrade";

Q. Is there a way to define this in the 'Custom Nginx Configuration' tab?

This is how it looks in a test nginx proxy standalone system:
---
server {
	listen 443 ssl;
	listen [::]:443 ssl;

	server_name ha.*;

	include /config/nginx/ssl.conf;

	client_max_body_size 0;

	# enable for ldap auth, fill in ldap details in ldap.conf
	#include /config/nginx/ldap.conf;

	location / {
		# enable the next two lines for http auth
		#auth_basic "Restricted";
		#auth_basic_user_file /config/nginx/.htpasswd;

		# enable the next two lines for ldap auth
		#auth_request /auth;
		#error_page 401 =200 /login;

		include /config/nginx/proxy.conf;
		resolver 127.0.0.1 valid=30s;
		set $upstream_homeassistant 192.168.1.22;
		proxy_pass http://$upstream_homeassistant:8123;

		proxy_set_header Upgrade $http_upgrade;
		proxy_set_header Connection "upgrade";
	}

	location /api/websocket {
		resolver 127.0.0.1 valid=30s;
		set $upstream_homeassistant 192.168.1.22;
		proxy_pass http://$upstream_homeassistant:8123;
		proxy_set_header Host $host;

		proxy_http_version 1.1;
		proxy_set_header Upgrade $http_upgrade;
		proxy_set_header Connection "upgrade";
	}
}
---

**Comment by cschwarck** [comment]
I found there is a setting in Home Assistant that you need to set allowing access through a proxy. Once set, access to HA now works correctly.

**Reply by cmitz** [reply]
Yes, this part of the config is required for a reverse proxy to work:
https://www.home-assistant.io/integrations/http/#trusted_proxies

**Comment by cmitz** [comment]
To answer your initial question: you can add a custom config inside a `location`, e.g. `/api/websocket` in this tab:
![afbeelding](https://user-images.githubusercontent.com/5594436/163794689-93296558-7e26-4226-8280-99ac455ecf57.png)

It should point to the same server and port, and when clicking on the cog-icon (aka gear-icon), you should be able to add those custom rules.

However, this is not recommended since you should just add the IP address of the proxy like you already found out
https://github.com/NginxProxyManager/nginx-proxy-manager/discussions/1401#discussioncomment-2585172

**Comment by Maharshi-Pathak** [comment]
I am getting an error "unable to connect to Home Assistant" after I updated NPM to latest version. Error: 
<img width="358" alt="image" src="https://github.com/NginxProxyManager/nginx-proxy-manager/assets/64644720/41d8c905-3bb2-4705-b1f7-af6a916ef22b">
## HA 
 This is the error in the HA log: 
```
2023-12-11 19:22:50.075 WARNING (MainThread) [homeassistant.components.http.ban] Login attempt or request with invalid authentication from IP-address
````
My configuration.yaml

```
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - server-gateway-ip
```
## NPM UI
SSL certificate config:
<img width="348" alt="image" src="https://github.com/NginxProxyManager/nginx-proxy-manager/assets/64644720/1b6fa1ab-c587-4cf6-8ffe-d9d8dbaed589">
Proxy host config
<img width="347" alt="image" src="https://github.com/NginxProxyManager/nginx-proxy-manager/assets/64644720/9fc5a748-73a2-4685-ae5b-325efcc583d7">


Any suggestions on how to resolve this?
